### PR TITLE
JS -> TS

### PR DESCRIPTION
**Pull request by anarast**
- converts all files from javascript to typescript, adds typescript eslint
- may have broken the prettier lint step

### CODE REVIEW

File: src/autoupdater.ts [hunk 1]
```diff
@@ -107,7 +114,7 @@ class AutoUpdater {
     return true;
   }
 
-  async prNeedsUpdate(pull) {
+  async prNeedsUpdate(pull: any): Promise<boolean> {
```
Comment (anarast) [inline src/autoupdater.ts:117]: I thought that this type should have been `PullsListResponseItem` like for `update()`, but using it gives the error `Property 'merged' does not exist on type 'PullsListResponseItem'`

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "compilerOptions": {
+    "types": ["node", "jest"],
+    "target": "es5",
```
Comment (chinthakagodawita) [inline tsconfig.json:4]: Could you target `ES2019`? See https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping - we run Node 12.x for now

File: test/autoupdate.test.ts [hunk 3]
```diff
@@ -53,8 +53,9 @@ describe('test `prNeedsUpdate`', () => {
   test('pull request has already been merged', async () => {
     const pull = {
       merged: true,
-    }
+    };
 
+    // const config = new ConfigLoader();
```
Comment (chinthakagodawita) [inline test/autoupdate.test.ts:58]: Could you remove this? There's a few more instances further down in this file

File: package.json [hunk 4]
```diff
@@ -10,28 +10,36 @@
     "autoupdate-action": "bin/cli.js"
   },
   "scripts": {
+    "build": "tsc && ncc build bin/cli.ts --out dist",
```
Comment (chinthakagodawita) [inline package.json:13]: Do we need `tsc`? I thought `ncc` supported Typescript out of the box: https://github.com/vercel/ncc#with-typescript

Comment (chinthakagodawita) [reply package.json:13]: Also - could you try a docker build and see if the `Dockerfile` still works?

File: .eslintrc [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (chinthakagodawita) [inline .eslintrc:1]: What's the reason for making this a JSON file instead of YAML? I'd prefer YAML tbh mostly because JSON doesn't allow for comments

File: package.json [hunk 6]
```diff
@@ -10,28 +10,36 @@
...
+    "argparse": "^1.0.10",
+    "typescript": "^3.9.7"
   },
   "devDependencies": {
-    "eslint": "^7.4.0",
+    "@types/jest": "^26.0.8",
+    "@typescript-eslint/eslint-plugin": "^3.7.1",
+    "@typescript-eslint/parser": "^3.7.1",
+    "eslint": "^7.6.0",
     "eslint-config-airbnb-base": "^14.2.0",
```
Comment (chinthakagodawita) [inline package.json:32]: Could you remove this if it's no longer used?

File: package.json [hunk 7]
```diff
@@ -10,28 +10,36 @@
...
     "test": "jest",
-    "test:watch": "jest --watchAll",
-    "build": "ncc build bin/cli.js --out dist"
+    "test:watch": "jest --watchAll"
   },
   "dependencies": {
     "@actions/core": "^1.2.4",
     "@actions/github": "^2.0.0",
+    "@octokit/types": "^5.1.0",
+    "@types/node": "^14.0.27",
```
Comment (chinthakagodawita) [inline package.json:22]: Still on node 12 for now

File: src/autoupdater.ts [hunk 8]
```diff
@@ -1,14 +1,20 @@
...
+import { ConfigLoader } from './config-loader';
 
-class AutoUpdater {
-  constructor(config, eventData) {
+export class AutoUpdater {
+  eventData: any;
+  config: ConfigLoader;
+  octokit: github.GitHub;
+
+  constructor(config: ConfigLoader, eventData: any) {
```
Comment (chinthakagodawita) [inline src/autoupdater.ts:11]: Is it possible to supress this check warning? Or typehint the correct type?

File: src/autoupdater.ts [hunk 9]
```diff
@@ -197,8 +204,8 @@ class AutoUpdater {
     return true;
   }
 
-  async merge(mergeOpts) {
-    const sleep = (timeMs) => {
+  async merge(mergeOpts: any): Promise<boolean> {
```
Comment (chinthakagodawita) [inline src/autoupdater.ts:207]: You should be able to typehint this, it's whatever `this.octokit.repos.merge` takes

File: src/autoupdater.ts (hunk 1 above)
Comment (chinthakagodawita) [reply src/autoupdater.ts:117]: Maybe a type out of https://github.com/octokit/webhooks is what you're looking for? This isn't sent via an API request, GH actions calls autoupdate with an pull event

File: src/config-loader.ts [hunk 10]
```diff
@@ -1,55 +1,57 @@
...
   }
 
-  mergeConflictAction() {
+  mergeConflictAction(): string {
     // one of 'fail' or 'ignore'.
     return this.getValue('MERGE_CONFLICT_ACTION', false, 'fail');
   }
 
-  getValue(key, required = false, defaulVal = null) {
+  getValue(key: string, required = false, defaultVal?: any): any {
```
Comment (chinthakagodawita) [inline src/config-loader.ts:54]: An `any` type makes sense here, could you disable this check for this function?

File: bin/cli.ts [hunk 11]
```diff
@@ -1,16 +1,16 @@
-#!/usr/bin/env node
-const fs = require('fs');
+#! node
```
Comment (chinthakagodawita) [inline bin/cli.ts:1]: Could you revert this change? `/usr/bin/env node` is much safer than calling `node` directly

File: package.json (hunk 4 above)
Comment (anarast) [reply package.json:13]: oh cool, yeah I'll remove it. yes the dockerfile seems to work still

**Comment by anarast** [comment]
exec'd into the container and the index.js looked like it had all the stuff
